src/visualizer.cpp

```diff
@@ -0,0 +1,522 @@
+#include <iostream>
+#include <CImg.h>
+#include <visualizer.h>
+#include <gate.h>
+
+using namespace cimg_library;
+using namespace ql;
+
+int main()
+{
+	// TODO: implement a generic grid structure object to contain the visual structure of the circuit, to ease positioning of components in all the drawing functions
+	// TODO: implement actual measurement symbol
+	// TODO: option to display the classical bit lines
+	// TODO: display wait/barrier
+	// TODO: gate duration outlines in gate color
+	// TODO: different types of cycle/duration(ns) labels
+	// TODO: 'cutting' circuits where nothing/not much is happening both in terms of idle cycles and idle qubits
+	// TODO: generate default gate visuals from the configuration file
+	// TODO: representing the gates as waveforms
+
+	Layout layout;
+
+	visualize(getTeleportationCircuit(), layout);
+	visualize(getDemoCircuit1(), layout);
+	visualize(getDemoCircuit2(), layout);
+
+	//layout.bitLine.groupClassicalLines = true;
+	//layout.cycles.compressCycles = true;
+
+	//visualize(getTeleportationCircuit(), layout);
+	//visualize(getDemoCircuit1(), layout);
+	//visualize(getDemoCircuit2(), layout);
+	
+	return 0;
+}
+
+void visualize(const std::vector<ql::gate*> gates, const Layout layout)
+{
+	validateLayout(layout);
+
+	// Calculate amount of cycles.
+	unsigned int amountOfCycles = 0;
+	for (const gate* gate : gates)
+	{
+		const unsigned int gateCycle = (unsigned int)gate->cycle;
+		if (gateCycle > amountOfCycles)
+			amountOfCycles = gateCycle;
+	}
+	amountOfCycles++; // because the cycles start at zero, we add one to get the true amount of cycles
```

Comment:
this should be in a private helper function

Reply:
Done